Radio schedule error logging

itimirichard opened the pull request:
Resolves #6272

**Overall change:**
_Logs errors when program fields are missing_

**Code changes:**

- _Added `logProgramError` utility function._
- _Check if any program is missing a required field or if program is incomplete and log the error._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [x] This PR requires manual testing

itimirichard committed:
log errors when required fields are missing

itimirichard committed:
Merge remote-tracking branch 'origin/latest' into radio-schedule-error-logging

tochwill commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -1,6 +1,10 @@
 import findLastIndex from 'ramda/src/findLastIndex';
 import propSatisfies from 'ramda/src/propSatisfies';
 import path from 'ramda/src/path';
+import webLogger from '#lib/logger.web';
```
Should we be using `nodeLogger from '#lib/logger.node';` instead here? You can then use the logger in this format:

logger.error(RADIO_SCHEDULE_DATA_INCOMPLETE_ERROR, { error });

I think webpack swaps out the node logger for the web logger when it is bundled for the client - this is going to be run on both!

tochwill replied:
I kept the web logger in the radio schedules component in the use effect, as it is also used in most read and the toggles container. A conversation for after this should we just be using one, standardised logger, and letting webpack handle the swap?

tochwill commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -51,6 +68,19 @@ export default (data, service, currentTime) => {
+        logMissingFieldError('title');
+      }
+      if (!episodeTitle) {
+        logMissingFieldError('presentationTitle');
```
I think we are interested in when we are unable to render any schedules as well, when `scheduleDataIsComplete` is false

sareh commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -51,6 +68,19 @@ export default (data, service, currentTime) => {
       } = program;
 
+      const brandTitle = path(['brand', 'title'], program);
+      const episodeTitle = path(['episode', 'presentationTitle'], program);
```
We now don't need to use episodeTitle at all, since version 2.0.0 of psammead-radio-schedules does not need it. 

https://github.com/bbc/psammead/pull/3377/files#diff-daa9b5cfc0c949bd0a8f361495e8a040L147

We can remove this from here.

itimirichard replied:
From the screenshot Laura shared, episodeTitle appears to be a minimum requirement. 
<img width="447" alt="Screen Shot 2020-04-22 at 14 12 16" src="https://user-images.githubusercontent.com/34133081/80005542-a201d000-84bb-11ea-8e98-97dea2c4dd5a.png">

tochwill replied:
We no longer pass in the episode title - it is formatted from the start time instead!

sareh commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -20,6 +24,19 @@ export const getLink = (state, program, service) => {
 };
 
+const logMissingFieldError = field => {
+  logger.error(
+    JSON.stringify(
+      {
+        event: RADIO_SCHEDULE_DATA_INCOMPLETE_ERROR,
+        message: `${field} field is missing in program`,
+      },
+      null,
+      2,
+    ),
+  );
+};
```
This can be updated to have a simpler format:
```suggestion
  logger.error({
      event: RADIO_SCHEDULE_DATA_INCOMPLETE_ERROR,
      message: `${field} field is missing in program`,
    }),
  );
};
```

sareh replied:
I believe you were looking at our logs in the file `src/app/containers/RadioSchedule/Canonical/index.jsx`, but these are currently using an outdated format that needs updating.

sareh reviewed (commented):
Thank you, @itimirichard for the quick PR!

I was testing this locally and I wasn't able to see this error in my console. This is the process I followed:
-  I updated the local data in `data/afrique/bbc_afrique_radio/schedule.json` to only have the last 4 items, and removed the brand.title from these remaining items. 
- I visited http://localhost:7080/afrique and checked my console. 

Can you please share how you got it working?

itimirichard committed:
use nodelogger, log error when scheduleData
is incomplete

tochwill commented:
> Thank you, @itimirichard for the quick PR!
> 
> I was testing this locally and I wasn't able to see this error in my console. This is the process I followed:
> 
> * I updated the local data in `data/afrique/bbc_afrique_radio/schedule.json` to only have the last 4 items, and removed the brand.title from these remaining items.
> * I visited http://localhost:7080/afrique and checked my console.
> 
> Can you please share how you got it working?

@sareh Are you looking at the server logs or web console? - For the server render these will only appear in server logs

itimirichard committed:
update logger function

itimirichard commented:
> Thank you, @itimirichard for the quick PR!
> 
> I was testing this locally and I wasn't able to see this error in my console. This is the process I followed:
> 
> * I updated the local data in `data/afrique/bbc_afrique_radio/schedule.json` to only have the last 4 items, and removed the brand.title from these remaining items.
> * I visited http://localhost:7080/afrique and checked my console.
> 
> Can you please share how you got it working?

I followed a similar process. However the erros are being logged to the app.log file not on the console

itimirichard committed:
remove episode title

itimirichard committed:
update tests

sareh reviewed (commented):
Thanks for the updates. Now that the node logger is used, I can see the errors in the `simorgh/log/app.log` file locally. 

<img width="599" alt="Screenshot logging" src="https://user-images.githubusercontent.com/3028997/80010472-60c0ee80-84c2-11ea-8df0-49332b47ae67.png">

sareh commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -51,6 +67,19 @@ export default (data, service, currentTime) => {
       } = program;
 
+      const brandTitle = path(['brand', 'title'], program);
+      const episodeTitle = path(['episode', 'presentationTitle'], program);
```
```suggestion
```

tochwill commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -20,6 +24,13 @@ export const getLink = (state, program, service) => {
 };
 
+const logProgramError = error => {
+  logger.error({
```
I think this was correct before, where you pass in the event name followed by an object containing the error - then it matches our sumologic log format

sareh replied:
Good point, I completely missed that.

itimirichard committed:
update radioScheduleShape after removing
episodeTitle

sareh commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -51,6 +67,19 @@ export default (data, service, currentTime) => {
+      if (!brandTitle) {
+        logProgramError('title field is missing in program');
+      }
+      if (!episodeTitle) {
+        logProgramError('episodeTitle field is missing in program');
+      }
```
```suggestion
```

sareh commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -20,6 +24,13 @@ export const getLink = (state, program, service) => {
 };
 
+const logProgramError = error => {
+  logger.error({
+    event: RADIO_SCHEDULE_DATA_INCOMPLETE_ERROR,
+    message: error,
+  });
```
```suggestion
  logger.error(RADIO_SCHEDULE_DATA_INCOMPLETE_ERROR, { error });
```
I think this format is correct. Apologies for my mistaken suggestion before!

itimirichard committed:
Update src/app/containers/RadioSchedule/utilities/processRadioSchedule.js

Co-Authored-By: Sareh <sareh@users.noreply.github.com>

itimirichard committed:
Merge branch 'latest' into radio-schedule-error-logging

sareh reviewed (commented):
In the case where there are no programs in the schedule data, I get this log:
<img width="594" alt="Screenshot 2020-04-22 at 18 10 28" src="https://user-images.githubusercontent.com/3028997/80012034-98309a80-84c4-11ea-86fb-83658e08d204.png">
and no radio schedule on the page, which is great. I'm thinking that it would be useful to also add `service` as well as `error` to the logged output, so we can see which service is affected.

In the case where we have incomplete data (removing brand.title / publishStartTime). It'd also be good to log which service it is so we can investigate.
<img width="588" alt="Screenshot logs with incomplete data" src="https://user-images.githubusercontent.com/3028997/80011894-69b2bf80-84c4-11ea-9115-f6e193c3b062.png">

It looks like this for me locally:

<img width="306" alt="Screenshot no timestamp and no brand title" src="https://user-images.githubusercontent.com/3028997/80012279-fa899b00-84c4-11ea-9c8b-d76da9930508.png">


I believe in these two error cases the program should not be rendered at all?

itimirichard committed:
add service to error logs

itimirichard committed:
Merge remote-tracking branch 'origin/latest' into radio-schedule-error-logging

itimirichard committed:
Merge branch radio-schedule-error-logging of https://github.com/bbc/simorgh into radio-schedule-error-logging

sareh reviewed (approved):
We discussed offline that this PR was only to add logging. I'm happy to keep the component behaviour as it is in this PR. Can you please write another Simorgh issue to investigate not rendering the component at all in the case of incomplete data?

sareh commented:
Merging in since it doesn't need to be tested locally by a Test Engineer.
